src/engraving/libmscore/note.cpp

```diff
@@ -2133,48 +2133,138 @@ void Note::setHeadHasParentheses(bool hasParentheses)
         }
         // if stems beside staff, do nothing
         else {
             return;
         }
-    } else {
-        onLine = !(line() & 1);
-    }
-
-    bool oddVoice = voice() & 1;
-    if (onLine) {
-        // displace dots by half spatium up or down according to voice
-        if (pos == DirectionV::AUTO) {
-            y = oddVoice ? 0.5 : -0.5;
-        } else if (pos == DirectionV::UP) {
-            y = -0.5;
+    } else if (onLine) {
+        // NON-TAB
+        bool oddVoice = voice() & 1;
+        Chord* c = chord();
+        bool hasVoices = c->measure()->hasVoices(c->staffIdx(), c->tick(), c->ticks());
+        std::vector<Note*> topDownNotes;
+        std::vector<Note*> bottomUpNotes;
+        std::vector<size_t> anchoredDots;
+        // construct combined chords using the notes from overlapping chords
+        if (!hasVoices) {
+            // only this voice, so topDownNotes is just the notes in the chord
+            for (Note* note : c->notes()) {
+                if (note->line() & 1) {
+                    int newOffset = 0;
+                    bool adjustDown = (c->voice() & 1) && !c->up();
+                    if (!anchoredDots.empty() && anchoredDots.back() == note->line()) {
+                        if (anchoredDots.size() >= 1 && anchoredDots[anchoredDots.size() - 2] == note->line() + adjustDown ? 2 : -2) {
+                            newOffset = adjustDown ? -2 : 2;
+                        } else {
+                            newOffset = adjustDown ? 2 : -2;
+                        }
+                    }
+                    anchoredDots.push_back(note->line() + newOffset);
+                } else {
+                    topDownNotes.push_back(note);
+                }
+            }
         } else {
-            y = 0.5;
+            // Get a list of notes in this staff that adjust dots from top down,
+            // bottom up, and also start our locked-in dot list by adding all lines where dots are
+            // guaranteed
+            Measure* m = c->measure();
+            size_t firstVoice = c->track() - c->voice();
+            for (size_t i = firstVoice; i < firstVoice + VOICES; ++i) {
+                if (Chord* voiceChord = m->findChord(c->tick(), i)) {
+                    /*if (voiceChord->dots() == 0) {
+                        continue;
+                    }*/
+                    bool startFromTop = !((voiceChord->voice() & 1) && !voiceChord->up());
+                    if (startFromTop) {
+                        for (Note* note : voiceChord->notes()) {
+                            if (note->line() & 1) {
+                                anchoredDots.push_back(note->line());
+                            } else {
+                                topDownNotes.push_back(note);
+                            }
+                        }
+                    } else {
+                        for (Note* note : voiceChord->notes()) {
+                            if (note->line() & 1) {
```

Comment:
Similar code (with & 1) is present in a lot of places. What does it mean?

Reply:
Odd/even check